framework-crates/objc2-metal/translation-config.toml

```diff
@@ -33,864 +33,213 @@ enum.MTLTensorDataType.use-value = true
-class.MTLRenderPipelineDescriptor.methods."setFragmentFunction:".unsafe = false
-class.MTLRenderPipelineDescriptor.methods.vertexDescriptor.unsafe = false
-class.MTLRenderPipelineDescriptor.methods."setVertexDescriptor:".unsafe = false
-class.MTLRenderPipelineDescriptor.methods.sampleCount.unsafe = false
-class.MTLRenderPipelineDescriptor.methods."setSampleCount:".unsafe = false
-class.MTLRenderPipelineDescriptor.methods.rasterSampleCount.unsafe = false
-class.MTLRenderPipelineDescriptor.methods."setRasterSampleCount:".unsafe = false
-class.MTLRenderPipelineDescriptor.methods.isAlphaToCoverageEnabled.unsafe = false
-class.MTLRenderPipelineDescriptor.methods."setAlphaToCoverageEnabled:".unsafe = false
-class.MTLRenderPipelineDescriptor.methods.isAlphaToOneEnabled.unsafe = false
-class.MTLRenderPipelineDescriptor.methods."setAlphaToOneEnabled:".unsafe = false
-class.MTLRenderPipelineDescriptor.methods.isRasterizationEnabled.unsafe = false
-class.MTLRenderPipelineDescriptor.methods."setRasterizationEnabled:".unsafe = false
-class.MTLRenderPipelineDescriptor.methods.maxVertexAmplificationCount.unsafe = false
-# Call supportsVertexAmplificationCount: on device first
-# setMaxVertexAmplificationCount:
-class.MTLRenderPipelineDescriptor.methods.colorAttachments.unsafe = false
-class.MTLRenderPipelineDescriptor.methods.depthAttachmentPixelFormat.unsafe = false
-class.MTLRenderPipelineDescriptor.methods."setDepthAttachmentPixelFormat:".unsafe = false
-class.MTLRenderPipelineDescriptor.methods.stencilAttachmentPixelFormat.unsafe = false
-class.MTLRenderPipelineDescriptor.methods."setStencilAttachmentPixelFormat:".unsafe = false
-class.MTLRenderPipelineDescriptor.methods.inputPrimitiveTopology.unsafe = false
-# Must be specified when layered rendering is enabled
-# setInputPrimitiveTopology:
-class.MTLRenderPipelineDescriptor.methods.tessellationPartitionMode.unsafe = false
-# Affects maxTessellationFactor
-# setTessellationPartitionMode:
-class.MTLRenderPipelineDescriptor.methods.maxTessellationFactor.unsafe = false
+protocol.MTLHeap.methods."setPurgeableState:".unsafe = false
+
+# Using the resource's contents in a memory-safe manner is very difficult
+# after this is called.
+protocol.MTLResource.methods.makeAliasable.unsafe = true
+
+# SAFETY: Modifying residency is safe, it's effectively the same as
+# controlling what's in the L1/L2/L3 cache on the CPU.
+# protocol.MTLResidencySet.methods.requestResidency.unsafe = false
+# protocol.MTLResidencySet.methods.endResidency.unsafe = false
+
+# TODO(breaking): Mark this as unsafe?
+class.MTLHeapDescriptor.methods."setType:".unsafe = false
+
+# These affect lifetime safety, and can cause use-after-free if used incorrectly.
+class.MTLCommandBufferDescriptor.methods."setRetainedReferences:".unsafe = true
+protocol.MTLCommandQueue.methods.commandBufferWithUnretainedReferences.unsafe = true
+protocol.MTLIOCommandQueue.methods.commandBufferWithUnretainedReferences.unsafe = true
```

Comment:
@MarijnS95 pointed out in https://github.com/madsmtm/objc2/pull/773#issuecomment-3352377509:

> * **Lifetime safety**: This [used to be tracked by Metal if not opted out](https://developer.apple.com/documentation/metal/mtlcommandqueue/makecommandbufferwithunretainedreferences()?language=objc) but that already wasn't possible when manually uploading `gpuReourceID()` values, and [is removed now in Metal 4](https://developer.apple.com/documentation/metal/understanding-the-metal-4-core-api?language=objc#Command-buffers).

Which is just.. Ugh! Not yet sure how we'll resolve that. Maybe all command buffer and command encoder methods need to be unsafe? Maybe it'll be fine since most of them already are?

Reply:
Yeah, all dispatches on the GPU (shader dispatches, draws, copy commands, ...) should likely be unsafe because of resource lifetimes :shrug:

Reply:
Re-reading the docs for [`commandBufferWithUnretainedReferences`](https://developer.apple.com/documentation/metal/mtlcommandqueue/makecommandbufferwithunretainedreferences()?language=objc), it says:
> Returns a command buffer from the command queue that doesn’t maintain strong references to resources.

So it sounds like the things that aren't retained only applies to `MTLResource` and subprotocols, and that it'd still keep a reference to any `MTLFence`, `MTLCommandBufferHandler`, `MTLSamplerState` etc. that you pass to the buffer / its encoders?

As such, I think we can again get around this by marking `MTLResource` as unsafe (and probably add another safety requirement to the docs)?